File: you_search.py
"""
You.com Search API Client
Provides a simple interface to query the You.com Search API
"""
import os
import requests
from typing import Dict, List, Optional, Any
from dotenv import load_dotenv

# Load environment variables
load_dotenv()


class YouSearchClient:
    """Client for interacting with You.com Search API"""

    BASE_URL = "https://api.ydc-index.io/v1/search"

    def __init__(self, api_key: Optional[str] = None):
        """
        Initialize the You.com Search API client

        Args:
            api_key: You.com API key. If not provided, will look for YOU_API_KEY env var
        """
        self.api_key = api_key or os.getenv("YOU_API_KEY")
        if not self.api_key:
            raise ValueError(
                "API key is required. Set YOU_API_KEY environment variable or pass api_key parameter"
            )

    def search(self, query: str, num_results: int = 10) -> Dict[str, Any]:
        """
        Perform a search query using You.com API

        Args:
            query: Search query string
            num_results: Number of results to return (default: 10)

        Returns:
            Dictionary containing search results with web and news data

        Raises:
            requests.exceptions.RequestException: If the API request fails
        """
        headers = {
            "X-API-Key": self.api_key
        }

        params = {
            "query": query
        }

        try:
            response = requests.get(
                self.BASE_URL,
                headers=headers,
                params=params,
                timeout=30
            )
            response.raise_for_status()
            return response.json()

        except requests.exceptions.RequestException as e:
            raise Exception(f"Error querying You.com API: {str(e)}")

    def format_results_for_llm(self, results: Dict[str, Any]) -> str:
        """
        Format search results in an LLM-friendly text format

        Args:
            results: Raw API response from You.com

        Returns:
            Formatted string optimized for LLM consumption
        """
        formatted = []

        # Add web results
        if "results" in results and "web" in results["results"]:
            web_results = results["results"]["web"]
            formatted.append("=== WEB SEARCH RESULTS ===\n")

            for idx, result in enumerate(web_results[:10], 1):
                title = result.get("title", "No title")
                url = result.get("url", "")
                description = result.get("description", "")
                snippets = result.get("snippets", [])

                formatted.append(f"{idx}. {title}")
                formatted.append(f"   URL: {url}")
                if description:
                    formatted.append(f"   Description: {description}")
                if snippets:
                    formatted.append(f"   Snippets:")
                    for snippet in snippets[:3]:  # Limit to 3 snippets per result
                        formatted.append(f"   - {snippet}")
                formatted.append("")

        # Add news results if available
        if "results" in results and "news" in results["results"]:
            news_results = results["results"]["news"]
            if news_results:
                formatted.append("\n=== NEWS RESULTS ===\n")

                for idx, result in enumerate(news_results[:5], 1):
                    title = result.get("title", "No title")
                    url = result.get("url", "")
                    description = result.get("description", "")

                    formatted.append(f"{idx}. {title}")
                    formatted.append(f"   URL: {url}")
                    if description:
                        formatted.append(f"   Description: {description}")
                    formatted.append("")

        return "\n".join(formatted)


def main():
    """Example usage of the You.com Search API client"""
    try:
        client = YouSearchClient()

        # Example search
        query = "latest developments in AI"
        print(f"Searching for: {query}\n")

        results = client.search(query)

        # Print formatted results
        formatted_output = client.format_results_for_llm(results)
        print(formatted_output)

        # Also print raw JSON for debugging
        print("\n=== RAW JSON RESPONSE ===")
        import json
        print(json.dumps(results, indent=2))

    except Exception as e:
        print(f"Error: {str(e)}")


if __name__ == "__main__":
    main()

File: lib/you-search.ts
/**
 * You.com Search API Client
 *
 * This template uses fetch directly so you can see exactly how the API works.
 * For production apps, you can also use the official SDK: @youdotcom-oss/sdk
 *
 * @see https://docs.you.com
 */

/**
 * Search options for filtering and pagination
 */
export interface SearchOptions {
  count?: number; // 1-100, default 10
  freshness?: 'day' | 'week' | 'month' | 'year' | string;
  offset?: number; // 0-9, for pagination
  country?: string; // ISO 3166 code (US, GB, DE, etc.)
  language?: string; // BCP 47 code (EN, ES, FR, etc.)
  safesearch?: 'off' | 'moderate' | 'strict';
}

/**
 * Individual search result item
 */
export interface SearchResult {
  title: string;
  url: string;
  description?: string;
  snippets?: string[];
  favicon_url?: string;
  age?: string; // Relative time like "2 hours ago", "1 day ago"
  page_age?: string; // ISO 8601 date from API
  thumbnail_url?: string; // For preview images
  contents?: { html?: string; markdown?: string };
  authors?: string[];
}

/**
 * Search API response structure containing web and news results
 */
export interface SearchResponse {
  results: {
    web?: SearchResult[];
    news?: SearchResult[];
  };
  metadata?: {
    search_uuid: string;
    query: string;
    latency: number;
  };
}

/**
 * Complete API response with query and results
 */
export interface SearchAPIResponse {
  query: string;
  results: {
    web?: SearchResult[];
    news?: SearchResult[];
  };
  metadata?: {
    search_uuid: string;
    query: string;
    latency: number;
  };
}

/**
 * Client for interacting with the You.com Search API
 */
export class YouSearchClient {
  private static readonly BASE_URL = 'https://ydc-index.io/v1/search';
  private apiKey: string;

  constructor(apiKey?: string) {
    this.apiKey = apiKey || process.env.YOU_API_KEY || '';
    if (!this.apiKey) {
      throw new Error('missing_api_key');
    }
  }

  /**
   * Searches the You.com API for the given query
   */
  async search(query: string, options?: SearchOptions): Promise<SearchResponse> {
    const headers = {
      'X-API-Key': this.apiKey,
    };

    const params = new URLSearchParams({ query });

    if (options) {
      if (options.count !== undefined) params.set('count', String(options.count));
      if (options.freshness) params.set('freshness', options.freshness);
      if (options.offset !== undefined) params.set('offset', String(options.offset));
      if (options.country) params.set('country', options.country);
      if (options.language) params.set('language', options.language);
      if (options.safesearch) params.set('safesearch', options.safesearch);
    }

    try {
      const response = await fetch(`${YouSearchClient.BASE_URL}?${params.toString()}`, {
        headers,
        next: { revalidate: 0 },
      });

      if (response.status === 401 || response.status === 403) {
        throw new Error(`Invalid API key (${response.status})`);
      }

      if (response.status === 429) {
        throw new Error(`Rate limit exceeded (${response.status})`);
      }

      if (!response.ok) {
        throw new Error(`API request failed: ${response.status} ${response.statusText}`);
      }

      return await response.json();
    } catch (error) {
      if (error instanceof Error && error.message.match(/^\w/)) {
        throw error;
      }
      throw new Error(
        `Error querying You.com API: ${error instanceof Error ? error.message : 'Unknown error'}`
      );
    }
  }

  /**
   * Formats search results into LLM-friendly text format
   */
  formatResultsForLLM(results: SearchResponse): string {
    const formatted: string[] = [];

    if (results.results?.web) {
      const webResults = results.results.web;
      formatted.push('=== WEB SEARCH RESULTS ===\n');

      webResults.slice(0, 10).forEach((result, idx) => {
        const title = result.title || 'No title';
        const url = result.url || '';
        const description = result.description || '';
        const snippets = result.snippets || [];

        formatted.push(`${idx + 1}. ${title}`);
        formatted.push(`   URL: ${url}`);
        if (description) {
          formatted.push(`   Description: ${description}`);
        }
        if (snippets.length > 0) {
          formatted.push(`   Snippets:`);
          snippets.slice(0, 3).forEach((snippet) => {
            formatted.push(`   - ${snippet}`);
          });
        }
        formatted.push('');
      });
    }

    if (results.results?.news && results.results.news.length > 0) {
      const newsResults = results.results.news;
      formatted.push('\n=== NEWS RESULTS ===\n');

      newsResults.slice(0, 5).forEach((result, idx) => {
        const title = result.title || 'No title';
        const url = result.url || '';
        const description = result.description || '';

        formatted.push(`${idx + 1}. ${title}`);
        formatted.push(`   URL: ${url}`);
        if (description) {
          formatted.push(`   Description: ${description}`);
        }
        formatted.push('');
      });
    }

    return formatted.join('\n');
  }
}
